src/test/java/com/emc/object/s3/S3JerseyClientTest.java

```diff
@@ -2971,4 +2972,172 @@ protected void assertAclEquals(AccessControlList acl1, AccessControlList acl2) {
+        List<ObjectTag> tagMarshalXml = new ArrayList<ObjectTag>(){{add(new ObjectTag("<k", "v%"));}};
+        // A tag key can be up to 128 Unicode characters in length, and tag values can be up to 256 Unicode characters in length
+        String largeKey = new Random().ints(48, 122 + 1).limit(128 + 1).collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append).toString();
+        String largeValue = new Random().ints(48, 122 + 1).limit(256 + 1).collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append).toString();
+        List<ObjectTag> tagLargeKV = new ArrayList<ObjectTag>(){{add(new ObjectTag(largeKey, largeValue));}};
+
+        // prepare PutObjectTaggingRequests
+        PutObjectTaggingRequest putObjectTaggingRequestSingleTag = new PutObjectTaggingRequest(bucketName, key).withTagging(new ObjectTagging().withTagSet(tag)),
+                putObjectTaggingRequestMultipleTags = new PutObjectTaggingRequest(bucketName, key).withTagging(new ObjectTagging().withTagSet(tags)),
+                putObjectTaggingRequestExceededTags = new PutObjectTaggingRequest(bucketName, key).withTagging(new ObjectTagging().withTagSet(tagsExceeded)),
+                putObjectTaggingRequestMarshalTag = new PutObjectTaggingRequest(bucketName, key).withTagging(new ObjectTagging().withTagSet(tagMarshalXml)),
+                putObjectTaggingRequestLargeKey = new PutObjectTaggingRequest(bucketName, key).withTagging(new ObjectTagging().withTagSet(tagLargeKV));
+
+        // get different versions of the existing object
+        client.putObject(bucketName, key, "Hello VersionID0 !", "text/plain");
+        String versionId = client.listVersions(bucketName, null).getVersions().get(0).getVersionId();
+
+        // GET the Tag of a non-existent object
+        try {
+            t = client.getObjectTagging(new GetObjectTaggingRequest(bucketName, "object-key-not-exist"));
+        } catch (S3Exception e) {
+            Assert.assertEquals(404, e.getHttpCode());
+            Assert.assertEquals("NoSuchKey", e.getErrorCode());
+        }
+
+        // should be able to update tags for a particular version of an object, 1 ~ 10
+        client.putObjectTagging(putObjectTaggingRequestSingleTag.withVersionId(versionId));
+        Assert.assertEquals(1, client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key).withVersionId(versionId)).getTagSet().size());
+        client.putObjectTagging(putObjectTaggingRequestMultipleTags.withVersionId(versionId));
+        Assert.assertEquals(10, client.getObjectTagging(new GetObjectTaggingRequest(bucketName, key).withVersionId(versionId)).getTagSet().size());
+
+        // should not be able to add more than 10 tags per object
+        try {
+            client.putObjectTagging(putObjectTaggingRequestExceededTags.withVersionId(versionId));
```

Comment:
Need to add Assert.fail("") to ensure the scenario of no exception is caught.
Please remember to enhance the similar issue in some other code places.

Reply:
have covered failure scenarios.